src/passes/OptimizeInstructions.cpp

```diff
@@ -1963,7 +1963,7 @@ struct OptimizeInstructions
     }
 
     // Check whether the cast will definitely succeed.
-    if (HeapType::isSubType(curr->ref->type.getHeapType(), intendedType)) {
+    if (Type::isSubType(curr->ref->type, curr->type)) {
```

Comment:
Instead of using `curr->ref->type` here, we could use the more refined of `cur->ref->type` and `fallthrough->type` for some extra precision in case type information was lost somewhere.

We can also check to see whether the cast will definitely fail by subsequently checking `!Type::isSubType(curr->type, curr->ref->type)`. It looks like we try to do that above with a call to `canBeCastTo`, but that only looks at heap types, so it's not as precise as it could be.

Reply:
Agreed, we can do more here. I suggest we leave those as followups?